### PR TITLE
"Partition pruning" for s3

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to prune the list of files via virtual columns such as `_file` and `_path` when reading from S3. This is for https://github.com/ClickHouse/ClickHouse/issues/37174 , https://github.com/ClickHouse/ClickHouse/issues/23494

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Functions/IFunction.cpp [hunk 1]
```diff
@@ -250,7 +250,10 @@ ColumnPtr IExecutableFunction::executeWithoutSparseColumns(const ColumnsWithType
                                         : columns_without_low_cardinality.front().column->size();
 
             auto res = executeWithoutLowCardinalityColumns(columns_without_low_cardinality, dictionary_type, new_input_rows_count, dry_run);
-            auto keys = res->convertToFullColumnIfConst();
+            bool res_is_constant = isColumnConst(*res);
```
Comment (amosbird) [inline src/Functions/IFunction.cpp:253]: When default implementation is used,  `Const(LowCardinality(...))` loses its constant property and becomes `LowCardinality(...)`, which in turn disables pruning with virtual columns.

Here is the fix.

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -684,12 +807,17 @@ bool StorageS3::isColumnOriented() const
...
+    SelectQueryInfo & query_info,
     ContextPtr local_context,
     QueryProcessingStage::Enum /*processed_stage*/,
     size_t max_block_size,
     unsigned num_streams)
 {
+    bool has_wildcards = s3_configuration.uri.bucket.find(PARTITION_ID_WILDCARD) != String::npos
+        || keys.back().find(PARTITION_ID_WILDCARD) != String::npos;
+    if (partition_by && has_wildcards)
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Reading from a partitioned S3 storage is not implemented yet");
```
Comment (amosbird) [inline src/Storages/StorageS3.cpp:819]: It's not possible to read from a partitioned S3 storage for now (it always returns empty). Let's throw an exception instead.

File: tests/queries/0_stateless/02302_s3_file_pruning.sql [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+select * from test_02302;  -- { serverError 48 }
+drop table test_02302;
+
+set max_rows_to_read = 1;
+
+-- Test s3 table function with glob
+select * from s3(s3_conn, filename='test_02302_*', format=Parquet) where _file like '%5';
+
+-- Test s3 table with explicit keys (no glob)
+-- TODO support truncate table function
```
Comment (amosbird) [inline tests/queries/0_stateless/02302_s3_file_pruning.sql:17]: In order to construct an S3 storage with explicit keys, I have to apply the following clusmy steps. Things can be improved:

1. allow to do `truncate table function s3(...)`
2. allow to create an S3 storage with a list of keys (or even different buckets)

**Review by kssenii — APPROVED** [review]
LGTM

**Pull request merged**